src/client/Client.cc

```diff
@@ -11274,28 +11284,32 @@ int64_t Client::_preadv_pwritev_locked(Fh *fh, const struct iovec *iov,
     }
 
     if (write) {
-        int64_t w = _write(fh, offset, totallen, NULL, iov, iovcnt, onfinish, do_fsync, syncdataonly);
+        int64_t w = _write(fh, offset, totallen, NULL, iov, iovcnt, onfinish,
+                           do_fsync, syncdataonly, zerocopy);
         ldout(cct, 3) << "pwritev(" << fh << ", \"...\", " << totallen << ", " << offset << ") = " << w << dendl;
         return w;
     } else {
         bufferlist bl;
+        ceph_assert(!zerocopy || blp);
         int64_t r = _read(fh, offset, totallen, blp ? blp : &bl,
                           onfinish);
         ldout(cct, 3) << "preadv(" << fh << ", " <<  offset << ") = " << r << dendl;
         if (r <= 0) {
           return r;
         }
 
-        client_lock.unlock();
-        copy_bufferlist_to_iovec(iov, iovcnt, blp ? blp : &bl, r);
-        client_lock.lock();
+	if (!zerocopy) {
+          client_lock.unlock();
+          copy_bufferlist_to_iovec(iov, iovcnt, blp ? blp : &bl, r);
+          client_lock.lock();
+        }
         return r;
     }
 }
 
 int Client::_preadv_pwritev(int fd, const struct iovec *iov, int iovcnt,
                             int64_t offset, bool write, Context *onfinish,
-                            bufferlist *blp)
+                            bufferlist *blp, bool zerocopy)
```

Comment:
`Client::_preadv_pwritev` is used by only two code paths - `Client::preadv` and `Client::pwritev`. You enabled zerocopy for `Client::_preadv_pwritev` but I guess the callers are not enabled yet to make use of zerocopy.

Reply:
Yea, I suppose I shouldn't have added zerocopy parameter to _preadv_pwritev since preadv and pwritev are not enabled for onfinish and blp either.

Reply:
Okay, so since you mentioned `onfinish`, do you mean zerocopy should only be used with nonblocking i/o?

Reply:
zero-copy can be used without non-blocking, it's just the preadv and pwritev don't use non-blocking. And really, adding non-blocking to them would change their APIs, so probably not worth adding unless a use case comes along.